### ISSUE TITLE
ondevice daemon: Connection timeout issue

### ISSUE THREAD

**Issue by mreithub** [ondevice/ondevice#14]
It seems like `ondevice daemon` doesn't have a connection timeout if there's no response from the API server.

Here's the stacktrace of a hanging `ondevice daemon` (the issue seems to be in `goroutine 1`):

```
SIGABRT: abort
PC=0x45a969 m=0

goroutine 0 [idle]:
runtime.epollwait(0x5, 0x7ffeb92e0c70, 0xffffffff00000080, 0xffffffff00000000, 0xffffffff00000001, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /usr/local/go/src/runtime/sys_linux_amd64.s:525 +0x19
runtime.netpoll(0x912001, 0x0)
        /usr/local/go/src/runtime/netpoll_epoll.go:67 +0x93
runtime.findrunnable(0xc420016000, 0x0)
        /usr/local/go/src/runtime/proc.go:2003 +0x28c
runtime.schedule()
        /usr/local/go/src/runtime/proc.go:2120 +0x14c
runtime.goexit0(0xc420001380)
        /usr/local/go/src/runtime/proc.go:2257 +0x180
runtime.mcall(0x7ffeb92e13f0)
        /usr/local/go/src/runtime/asm_amd64.s:240 +0x5b

goroutine 1 [IO wait, 313 minutes]:
net.runtime_pollWait(0x7f6d118d5d80, 0x72, 0x6)
        /usr/local/go/src/runtime/netpoll.go:160 +0x59
net.(*pollDesc).wait(0xc420059790, 0x72, 0xc420042db0, 0xc420062080)
        /usr/local/go/src/net/fd_poll_runtime.go:73 +0x38
net.(*pollDesc).waitRead(0xc420059790, 0x8f2020, 0xc420062080)
        /usr/local/go/src/net/fd_poll_runtime.go:78 +0x34
net.(*netFD).Read(0xc420059730, 0xc420025400, 0x400, 0x400, 0x0, 0x8f2020, 0xc420062080)
        /usr/local/go/src/net/fd_unix.go:243 +0x1a1
net.(*conn).Read(0xc420022018, 0xc420025400, 0x400, 0x400, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/net.go:173 +0x70
crypto/tls.(*block).readFromUntil(0xc420012360, 0x7f6d118d5fe8, 0xc420022018, 0x5, 0xc420022018, 0xc4200001a0)
        /usr/local/go/src/crypto/tls/conn.go:476 +0x91
crypto/tls.(*Conn).readRecord(0xc42010c000, 0x416, 0x9c, 0x0)
        /usr/local/go/src/crypto/tls/conn.go:578 +0xc4
crypto/tls.(*Conn).readHandshake(0xc42010c000, 0xc420110016, 0xc420110000, 0x97, 0x97)
        /usr/local/go/src/crypto/tls/conn.go:929 +0x8f
crypto/tls.(*Conn).clientHandshake(0xc42010c000, 0x7b0870, 0xc42010c108)
        /usr/local/go/src/crypto/tls/handshake_client.go:159 +0x99f
crypto/tls.(*Conn).Handshake(0xc42010c000, 0x0, 0x0)
        /usr/local/go/src/crypto/tls/conn.go:1260 +0x1b8
github.com/gorilla/websocket.(*Dialer).Dial(0x90e1e0, 0xc420061080, 0x37, 0xc4200e48d0, 0x0, 0x0, 0x0, 0x0)
        /tmp/build/src/github.com/gorilla/websocket/client.go:331 +0x15c4
github.com/ondevice/ondevice/tunnel.OpenWebsocket(0xc420085540, 0x780d91, 0x6, 0xc4200e48a0, 0xc4200f80e0, 0xc420043c18, 0x1, 0x1, 0xc420060e00, 0x40)
        /tmp/build/src/github.com/ondevice/ondevice/tunnel/ws.go:47 +0x411
github.com/ondevice/ondevice/daemon.Connect(0x0, 0x1, 0x1, 0x1, 0x1, 0xc4200cd4a0)
        /tmp/build/src/github.com/ondevice/ondevice/daemon/daemon.go:47 +0x218
github.com/ondevice/ondevice/command.(*DaemonCommand).run(0x92e6b0, 0xc42005c070, 0x3, 0x3, 0xc420096988)
        /tmp/build/src/github.com/ondevice/ondevice/command/daemon.go:101 +0x122
github.com/ondevice/ondevice/command.Run(0x7ffeb92e2ea8, 0x6, 0xc42005c070, 0x3, 0x3, 0x8f4fe0)
        /tmp/build/src/github.com/ondevice/ondevice/command/command.go:49 +0xb3
main.main()
        /tmp/build/src/github.com/ondevice/ondevice/ondevice.go:17 +0x92

goroutine 17 [syscall, 313 minutes, locked to thread]:
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:2086 +0x1

goroutine 20 [IO wait]:
net.runtime_pollWait(0x7f6d118d5f00, 0x72, 0x0)
        /usr/local/go/src/runtime/netpoll.go:160 +0x59
net.(*pollDesc).wait(0xc4200d6060, 0x72, 0xc420033d10, 0xc420062080)
        /usr/local/go/src/net/fd_poll_runtime.go:73 +0x38
net.(*pollDesc).waitRead(0xc4200d6060, 0x8f2020, 0xc420062080)
        /usr/local/go/src/net/fd_poll_runtime.go:78 +0x34
net.(*netFD).accept(0xc4200d6000, 0x0, 0x8f0b20, 0xc4200cc960)
        /usr/local/go/src/net/fd_unix.go:419 +0x238
net.(*UnixListener).accept(0xc420010180, 0x4328eb, 0xc420033dd0, 0x5be1dd)
        /usr/local/go/src/net/unixsock_posix.go:158 +0x32
net.(*UnixListener).Accept(0xc420010180, 0x7b01d8, 0xc420094480, 0x8f4d20, 0xc4200120f0)
        /usr/local/go/src/net/unixsock.go:229 +0x49
net/http.(*Server).Serve(0xc4200d8100, 0x8f44a0, 0xc420010180, 0x0, 0x0)
        /usr/local/go/src/net/http/server.go:2273 +0x1ce
net/http.Serve(0x8f44a0, 0xc420010180, 0x0, 0x0, 0x0, 0xc420010180)
        /usr/local/go/src/net/http/server.go:2091 +0x99
github.com/ondevice/ondevice/control.(*ControlSocket).run(0xc42006a050, 0x78097a, 0x4, 0x7ffeb92e2f06, 0x1f)
        /tmp/build/src/github.com/ondevice/ondevice/control/server.go:55 +0x131
created by github.com/ondevice/ondevice/control.StartServer
        /tmp/build/src/github.com/ondevice/ondevice/control/server.go:34 +0xe0
```

Look if there's a timeout field or something provided by gorilla websockets or add a timer to close the websocket if no server response was received in time.

To test this it should be sufficient to either add a `/etc/hosts` entry pointing at a firewalled IP or to set up the local firewall to drop requests to the API server's IP.

**Comment by mreithub** [comment]
Ok, blocking all traffic to the API server doesn't do the trick (go's socket Dial will timeout) but pointing `api.ondevice.io` to `127.0.0.1` using `/etc/hosts` and listen there using `sudo netcat -l -p 443` (-> missing SSL handshake) seems to be one way to trigger this bug.

**Comment by mreithub** [comment]
should've been fixed by 2f57cbbb
